### PR TITLE
Sentry SDK Crash Detection POC

### LINKED ISSUE #49749: Sentry SDK Crash Detection POC
### Problem Statement

This issue's aim is to enable the Sentry SDK Crash Detection (https://github.com/getsentry/sentry/issues/44342) only for the [sentry-sdk](https://sentry-sdks.sentry.io/) organization for a proof of concept.

### Solution Brainstorm

```[tasklist]
### Open PRs / Issues
- [ ] https://github.com/getsentry/sentry/pull/49170
- [ ] https://github.com/getsentry/sentry/pull/49528
- [ ] https://github.com/getsentry/sentry/pull/49593
- [ ] https://github.com/getsentry/sentry/pull/49746
- [ ] https://github.com/getsentry/sentry/pull/49747
- [ ] https://github.com/getsentry/sentry/issues/49935
- [ ] https://github.com/getsentry/sentry/issues/49938
- [ ] https://github.com/getsentry/sentry/issues/50710
- [ ] https://github.com/getsentry/sentry/issues/50712
- [ ] https://github.com/getsentry/getsentry/pull/11005
```

```[tasklist]
### PRs for splitting up the big POC PR
- [ ] https://github.com/getsentry/sentry/pull/50794
- [ ] https://github.com/getsentry/sentry/pull/50796
- [ ] https://github.com/getsentry/sentry/pull/51021
- [ ] https://github.com/getsentry/sentry/pull/51022
``` 

After merging the open PRs, it would be great to get a final review by @getsentry/ingest on the [diff](https://github.com/getsentry/sentry/compare/master...feat/sdk-crash-monitoring?expand=1) of the integration branch `feat/sdk-crash-monitoring`. Furthermore, I will validate that the feature is as expected by testing it locally before opening the PR. It's worth noting that all changes are kept behind the feature `organizations:sdk-crash-reporting` which is disabled by default.

PR (https://github.com/getsentry/getsentry/pull/11005) for enabling the feature in production just for the [sentry-sdks](https://sentry-sdks.sentry.io/)

### PR DESCRIPTION
**Pull request by philippsuess**
This PR adds the Sentry SDK Crash Detection (https://github.com/getsentry/sentry/issues/44342) only for the [sentry-sdk](https://sentry-sdks.sentry.io/) organization for a proof of concept. The feature is disabled by default.

Most of the code was already reviewed in previous PRs (https://github.com/getsentry/sentry/pull/46494, https://github.com/getsentry/sentry/pull/46578, https://github.com/getsentry/sentry/pull/49170, https://github.com/getsentry/sentry/pull/49528, https://github.com/getsentry/sentry/pull/49593, https://github.com/getsentry/sentry/pull/49746, https://github.com/getsentry/sentry/pull/50284)  but giving it a final pass still makes sense. 

Fixes GH-49749

### CODE REVIEW

**Review by iker-barriocanal — COMMENTED** [review]
Is there any chance we could split this PR into smaller ones? It's a bit hard to do a meaningful review on >1k LOC.

**Comment by philipphofmann** [comment]
> Most of the code was already reviewed in previous PRs, but giving it a final pass still makes sense.

@iker-barriocanal, ⬆️.

File: tests/sentry/utils/sdk_crashes/test_sdk_crash_detection.py [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+@region_silo_test
+class SDKCrashDetectionTest(
+    TestCase,
+    CococaSDKTestMixin,
+    CococaSDKFunctionTestMixin,
+    CococaSDKFilenameTestMixin,
+    CococaSDKFramesTestMixin,
+    PerformanceEventTestMixin,
+    SDKCrashReportTestMixin,
+):
```
Comment (asottile-sentry) [inline tests/sentry/utils/sdk_crashes/test_sdk_crash_detection.py:333]: please avoid writing new tests as class tests

File: tests/sentry/utils/sdk_crashes/test_fixture.py [hunk 2]
```diff
@@ -0,0 +1,269 @@
+from typing import Any, Mapping, Sequence
```
Comment (asottile-sentry) [inline tests/sentry/utils/sdk_crashes/test_fixture.py:1]: fixtures belong in the `fixtures/` namespace.  the `tests/` namespace is for tests

File: tests/sentry/utils/sdk_crashes/test_event_stripper.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+            )
+            == 0
+        ), "in_app frame should be removed"
+
+
+@region_silo_test
+class EventStripperTest(
+    TestCase,
+    EventStripperTestMixin,
+):
```
Comment (asottile-sentry) [inline tests/sentry/utils/sdk_crashes/test_event_stripper.py:136]: this "write a bunch of mixins" pattern is very strange -- I would not recommend it (where did you pick up this pattern?)

the recommended form of tests for pytest is to write function tests -- you can have many functions.   you can also write many classes (you don't need just one) but it is not recommended to write class-based tests

File: src/sentry/utils/sdk_crashes/cocoa_sdk_crash_detector.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
+from typing import Any, Mapping, Sequence
+
+from sentry.utils.glob import glob_match
+from sentry.utils.sdk_crashes.sdk_crash_detector import SDKCrashDetector
+
+
+class CocoaSDKCrashDetector(SDKCrashDetector):
+    def __init__(self) -> None:
+        self
```
Comment (asottile-sentry) [inline src/sentry/utils/sdk_crashes/cocoa_sdk_crash_detector.py:9]: this doesn't do anything

File: src/sentry/utils/sdk_crashes/cocoa_sdk_crash_detector.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    def is_sdk_frame(self, frame: Mapping[str, Any]) -> bool:
+        function = frame.get("function")
+
+        if function is not None:
+            # [SentrySDK crash] is a testing function causing a crash.
+            # Therefore, we don't want to mark it a as a SDK crash.
+            if "SentrySDK crash" in function:
+                return False
+
+            functionsMatchers = ["*sentrycrash*", "**[[]Sentry*"]
```
Comment (asottile-sentry) [inline src/sentry/utils/sdk_crashes/cocoa_sdk_crash_detector.py:36]: python uses `snake_case`

File: src/sentry/utils/sdk_crashes/event_stripper.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
+from typing import Any, Mapping, Sequence, Tuple
+
+from sentry.eventstore.models import Event
+from sentry.utils.safe import get_path
+from sentry.utils.sdk_crashes.sdk_crash_detector import SDKCrashDetector
+
+ALLOWED_EVENT_KEYS = {
```
Comment (asottile-sentry) [inline src/sentry/utils/sdk_crashes/event_stripper.py:7]: this should be a frozenset

File: src/sentry/utils/sdk_crashes/sdk_crash_detection.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+from sentry.issues.grouptype import GroupCategory
+from sentry.utils.safe import get_path, set_path
+from sentry.utils.sdk_crashes.cocoa_sdk_crash_detector import CocoaSDKCrashDetector
+from sentry.utils.sdk_crashes.event_stripper import strip_event_data
+from sentry.utils.sdk_crashes.sdk_crash_detector import SDKCrashDetector
+
+
+class SDKCrashReporter:
+    def __init__(self) -> None:
+        self
```
Comment (asottile-sentry) [inline src/sentry/utils/sdk_crashes/sdk_crash_detection.py:17]: ""

**Review by asottile-sentry — COMMENTED** [review]
this patch is unreviewably large -- it should be split into smaller patches

File: src/sentry/conf/server.py [hunk 8]
```diff
@@ -1564,6 +1564,8 @@ def SOCIAL_AUTH_DEFAULT_USERNAME():
     "organizations:slack-escape-messages": False,
     # If true, allow to create/use org auth tokens
     "organizations:org-auth-tokens": False,
+    # Enable detecting SDK crashes during event processing
```
Comment (philipphofmann) [inline src/sentry/conf/server.py:1567]: Code was already reviewed in https://github.com/getsentry/sentry/pull/49528

File: src/sentry/tasks/post_process.py [hunk 9]
```diff
@@ -1042,6 +1042,28 @@ def fire_error_processed(job: PostProcessJob):
         )
 
 
+def sdk_crash_monitoring(job: PostProcessJob):
```
Comment (philipphofmann) [inline src/sentry/tasks/post_process.py:1045]: Code was reviewed in https://github.com/getsentry/sentry/pull/49170

File: src/sentry/features/__init__.py [hunk 10]
```diff
@@ -178,6 +178,7 @@
 default_manager.add("organizations:set-grouping-config", OrganizationFeature, FeatureHandlerStrategy.INTERNAL)
 default_manager.add("organizations:slack-escape-messages", OrganizationFeature, FeatureHandlerStrategy.INTERNAL)
 default_manager.add("organizations:slack-overage-notifications", OrganizationFeature, FeatureHandlerStrategy.REMOTE)
+default_manager.add("organizations:sdk-crash-detection", OrganizationFeature, FeatureHandlerStrategy.INTERNAL)
```
Comment (philipphofmann) [inline src/sentry/features/__init__.py:181]: Code was reviewed in https://github.com/getsentry/sentry/pull/49528

File: src/sentry/utils/sdk_crashes/sdk_crash_detection.py [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+
+    def report(self, event_data: Dict[str, Any], event_project_id: int) -> Event:
+        from sentry.event_manager import EventManager
+
+        manager = EventManager(event_data)
+        manager.normalize()
+        return manager.save(project_id=event_project_id)
```
Comment (philipphofmann) [inline src/sentry/utils/sdk_crashes/sdk_crash_detection.py:24]: Code was reviewed in https://github.com/getsentry/sentry/pull/49593

File: tests/sentry/tasks/test_post_process.py [hunk 12]
```diff
@@ -1482,6 +1483,61 @@ def test_forecast_in_activity(self, mock_is_escalating):
         ).exists()
 
 
+class SDKCrashMonitoringTestMixin(BasePostProgressGroupMixin):
```
Comment (philipphofmann) [inline tests/sentry/tasks/test_post_process.py:1486]: Code was reviewed in https://github.com/getsentry/sentry/pull/49170

File: tests/sentry/utils/sdk_crashes/test_sdk_crash_detection.py (hunk 1 above)
Comment (philipphofmann) [reply tests/sentry/utils/sdk_crashes/test_sdk_crash_detection.py:333]: Thanks for pointing that out. I already created an issue for investigating which tests I can convert to function tests: https://github.com/getsentry/sentry/issues/50061

**Review by philippsuess — COMMENTED** [review]
I apologize for the large diff @iker-barriocanal and @asottile-sentry. As I pointed out in the PR description, most of the code was already reviewed in previous PRs, so I added some comments to let you know which parts you can skip. All the code in this PR is hidden behind a feature flag that is disabled by default. The goal is to only turn the feature on for the sentry-sdk organization. Therefore, I don't need a full detailed review focusing on each line of code. Instead, a high-level review to ensure the feature makes sense and won't cause trouble in production is what I would need. Would you still like me to split the PR up into multiple ones?

File: tests/sentry/utils/sdk_crashes/test_event_stripper.py (hunk 3 above)
Comment (philipphofmann) [reply tests/sentry/utils/sdk_crashes/test_event_stripper.py:136]: Initially, I had function-based tests. When I put it into [post_process.py](https://github.com/getsentry/sentry/blob/b953b0911253e786d8e4541b6f8a08feeab167fb/src/sentry/tasks/post_process.py) it didn't work correctly. So I changed the code to accept events as they are passed around in post_process.py. Then I looked at the [test_post_process.py](https://github.com/getsentry/sentry/blob/b953b0911253e786d8e4541b6f8a08feeab167fb/tests/sentry/tasks/test_post_process.py) and copied the pattern to make it work. If you have any suggestions to improve this, please let me know. I'm also not a fan of this pattern.

File: src/sentry/utils/sdk_crashes/cocoa_sdk_crash_detector.py (hunk 5 above)
Comment (philipphofmann) [reply src/sentry/utils/sdk_crashes/cocoa_sdk_crash_detector.py:33]: Thanks for pointing that out. Still getting used to Python. I'm a CamelCase 🐪 engineer.

File: src/sentry/utils/sdk_crashes/event_stripper.py (hunk 6 above)
Comment (philipphofmann) [reply src/sentry/utils/sdk_crashes/event_stripper.py:7]: TIL 😃

File: src/sentry/utils/sdk_crashes/event_stripper.py [hunk 13]
```diff
@@ -0,0 +1,130 @@
+from enum import Enum, auto
+from typing import Any, Dict, Mapping, Optional, Sequence
+
+from sentry.eventstore.models import Event
+from sentry.utils.safe import get_path
+from sentry.utils.sdk_crashes.sdk_crash_detector import SDKCrashDetector
+
+
+class Allow(Enum):
```
Comment (philipphofmann) [inline src/sentry/utils/sdk_crashes/event_stripper.py:9]: Code of this file was reviewed in https://github.com/getsentry/sentry/pull/50400.

**Review by jjbayer — CHANGES_REQUESTED** [review]
Pressing "Request changes" because I am still worried about user-specific data slipping through the cracks because of the `Allow.ALL` variant. I should have made that concern more clear when reviewing https://github.com/getsentry/sentry/pull/50400#pullrequestreview-1464631921.

Even if the feature is disabled by default, I am afraid that if we merge it now we might overlook this risk when opting in more orgs.

**Comment by philipphofmann** [comment]
> Pressing "Request changes" because I am still worried about user-specific data slipping through the cracks because of the `Allow.ALL` variant. I should have made that concern more clear when reviewing [#50400 (review)](https://github.com/getsentry/sentry/pull/50400#pullrequestreview-1464631921).
> 
> Even if the feature is disabled by default, I am afraid that if we merge it now we might overlook this risk when opting in more orgs.

@jjbayer, I get your point. I added https://github.com/getsentry/sentry/issues/50710 and https://github.com/getsentry/sentry/issues/50712 to https://github.com/getsentry/sentry/issues/49749. I will open PRs to this PR soon.

**Comment by philipphofmann** [comment]
I agree, @asottile-sentry. I should have started with a feature flag and then merge all the small batches into mainline. Smaller batches are better for tons of reasons.

@jjbayer, can we maybe agree on merging this PR and not enabling the feature until we address https://github.com/getsentry/sentry/issues/50710 and https://github.com/getsentry/sentry/issues/50712. This PR is already huge.

**Comment by asottile-sentry** [comment]
> I agree, @asottile-sentry. I should have started with a feature flag and then merge all the small batches into mainline. Smaller batches are better for tons of reasons.
> 
> @jjbayer, can we maybe agree on merging this PR and not enabling the feature until we address #50710 and #50712. This PR is already huge.

I think you should still try and split this up -- merging large unreviewable changes is inherently risky

**Comment by philipphofmann** [comment]
> I think you should still try and split this up -- merging large unreviewable changes is inherently risky

Okay, I will split this PR up into multiple ones. It shouldn't be a big problem. I just wanted to be sure to align on the approach. @asottile-sentry, I plan on adding a feature flag first without any actual logic. The feature will be disabled by default. Then I open subsequent PRs until all of the changes from this PR are merged to master. Only when we are happy will all the changes, we enable the feature. How does that sound @asottile-sentry?

**Comment by philipphofmann** [comment]
As this PR is too big, I will split it into multiple ones to make it easier to review.

**Pull request closed without merging**